apollo-federation/src/sources/connect/json_selection/js_literal.rs

```diff
@@ -0,0 +1,185 @@
+use nom::sequence::preceded;
+use nom::sequence::tuple;
+use nom::IResult;
+use serde::ser::SerializeMap;
+use serde::ser::SerializeSeq;
+use serde::Serialize;
+use serde_json_bytes::Value as JSON;
+
+use super::helpers::spaces_or_comments;
+use super::parser::parse_string_literal;
+use super::parser::Key;
+use super::parser::PathSelection;
+
+#[derive(Debug, PartialEq, Clone)]
+pub enum JSLiteral {
+    String(String),
+    Number(String),
+    Bool(bool),
+    Null,
+    Object(IndexMap<String, JSLiteral>),
+    Array(Vec<JSLiteral>),
+    Path(PathSelection),
+}
+
+impl JSLiteral {
+    // JSLiteral   ::= JSPrimitive | JSObject | JSArray | PathSelection
+    // JSPrimitive ::= StringLiteral | JSNumber | "true" | "false" | "null"
+    pub fn parse(input: &str) -> IResult<&str, Self> {
+        tuple((
+            spaces_or_comments,
+            alt((
+                map(parse_string_literal, Self::String),
+                Self::parse_number,
+                map(tag("true"), |_| Self::Bool(true)),
+                map(tag("false"), |_| Self::Bool(false)),
+                map(tag("null"), |_| Self::Null),
+                Self::parse_object,
+                Self::parse_array,
+                // TODO Disallow KeyPath PathSelection in JSLiteral?
+                map(PathSelection::parse, Self::Path),
+            )),
+            spaces_or_comments,
+        ))(input)
+        .map(|(input, (_, value, _))| (input, value))
+    }
+
+    // JSNumber    ::= "-"? (UnsignedInt ("." [0-9]*)? | "." [0-9]+)
+    // UnsignedInt ::= "0" | [1-9] NO_SPACE [0-9]*
+    fn parse_number(input: &str) -> IResult<&str, Self> {
+        delimited(
+            spaces_or_comments,
+            tuple((
+                opt(char('-')),
+                spaces_or_comments,
+                tuple((
+                    recognize(alt((
+                        recognize(char('0')),
+                        recognize(pair(one_of("123456789"), many0(one_of("0123456789")))),
+                    ))),
+                    opt(preceded(char('.'), recognize(many0(one_of("0123456789"))))),
+                )),
+            )),
+            spaces_or_comments,
+        )(input)
+        .map(|(input, (neg, _, (integer, decimal)))| {
+            let mut number = String::new();
+            if let Some(neg) = neg {
+                number.push(neg);
+            }
+            number.push_str(integer);
+            if let Some(decimal) = decimal {
+                number.push('.');
+                number.push_str(decimal);
+            }
+            (input, Self::Number(number))
+        })
+    }
+
+    // JSObject ::= "{" (JSProperty ("," JSProperty)*)? "}"
+    fn parse_object(input: &str) -> IResult<&str, Self> {
+        delimited(
+            tuple((spaces_or_comments, char('{'), spaces_or_comments)),
+            map(
+                opt(tuple((
+                    Self::parse_property,
+                    many0(preceded(char(','), Self::parse_property)),
```

Comment:
We should allow a trailing comma as well—which is something I stylistically _enforce_ in other languages to keep Git diffs cleaner.

Reply:
Good idea, and nice that this only affects the parser and nothing else thereafter: https://github.com/apollographql/router/pull/5762/commits/491232bda321cf35c56e77ebb8ba46b9409351f3